File: BasicDataStructure/Hash/twoSum.py
from typing import List

class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        # 记录“某个数值->它出现过的下标”
        seen = {}
        for i, x in enumerate(nums):
            need = target - x
            if need in seen:
                return [seen[need],i]
            seen[x] = i
        return []
"""
输入：nums = [2,7,11,15], target = 9
输出：[0,1]

T: O(n)
S: O(n)

"""

File: BasicDataStructure/Hash/twoSum.cpp
#include <vector>      // 提供vector容器
#include <unordered_map> // 提供unordered_map容器
using namespace std;    // 使用std命名空间
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {

        // 哈希表：key = 数值, value = 数值对应的下标
        unordered_map<int,int> seen;

        // 遍历数组
        for (int i = 0; i < (int)nums.size(); ++i){

            // 计算当前数字需要的配对值
            int need = target - nums[i];

            // 在哈希表中查找 need 是否已经出现过
            if (seen.count(need)) {
                // 如果出现过，说明找到了答案
                return {seen[need], i};
            }

            // 如果没有找到，把当前数字存入哈希表
            // 记录该数字出现的位置
            seen[nums[i]] = i;
        }

        // 题目保证一定有解，这里只是语法需要
        // return {};
    }
};